File: python/src/code_smells/change_preventers/shotgun_surgery.py
"""
Shotgun Surgery Code Smell Example

A single change requires making small modifications across many different classes.
When you need to change the tax rate, you must modify multiple classes.

Refactoring suggestion: Move Method, Move Field, Inline Class
"""
from typing import TypedDict


class LineItem(TypedDict):
    name: str
    price: float
    qty: int


class PriceCalculator:
    """Tax calculation logic duplicated here"""

    def total_with_tax(self, items: list[LineItem]) -> float:
        subtotal = sum(item['price'] * item['qty'] for item in items)
        tax = subtotal * 0.21
        return subtotal + tax


class InvoiceService:
    """Tax calculation logic duplicated here"""

    def create_total(self, items: list[LineItem]) -> float:
        base = sum(item['price'] * item['qty'] for item in items)
        vat = base * 0.21
        return base + vat


class SalesReport:
    """Tax calculation logic duplicated here"""

    def summarize(self, items: list[LineItem]) -> str:
        total_sum = sum(item['price'] * item['qty'] for item in items)
        tax = total_sum * 0.21
        total = total_sum + tax
        return f"total={total:.2f}"


class LoyaltyPoints:
    """Tax calculation logic duplicated here"""

    def points(self, items: list[LineItem]) -> int:
        base = sum(item['price'] * item['qty'] for item in items)
        with_tax = base + base * 0.21
        return int(with_tax // 10)


def demo_shotgun(items: list[LineItem]) -> tuple[float, float]:
    calc = PriceCalculator().total_with_tax(items)
    inv = InvoiceService().create_total(items)
    return (calc, inv)

File: go/code_smells/change_preventers/shotgun_surgery.go
package change_preventers

import "fmt"

// Code smell: Shotgun Surgery [Cirugía de escopeta].
// La misma regla de impuestos está duplicada en muchas clases;
// cambiarla requiere ediciones en múltiples lugares.

// Ejercicio: Cambia el impuesto del 21% al 18.5% con redondeo a 2 decimales.

// Tendrás que buscar cada copia y asegurar un redondeo consistente en todas partes,
// destacando cómo la duplicación convierte un cambio pequeño en muchas ediciones arriesgadas.

type LineItem struct {
	Name  string
	Price float64
	Qty   int
}

type PriceCalculator struct{}

func (p *PriceCalculator) TotalWithTax(items []LineItem) float64 {
	subtotal := 0.0
	for _, item := range items {
		subtotal += item.Price * float64(item.Qty)
	}
	tax := subtotal * 0.21
	return subtotal + tax
}

type InvoiceService struct{}

func (i *InvoiceService) CreateTotal(items []LineItem) float64 {
	base := 0.0
	for _, item := range items {
		base += item.Price * float64(item.Qty)
	}
	vat := base * 0.21
	return base + vat
}

type SalesReport struct{}

func (s *SalesReport) Summarize(items []LineItem) string {
	sum := 0.0
	for _, item := range items {
		sum += item.Price * float64(item.Qty)
	}
	tax := sum * 0.21
	total := sum + tax
	return fmt.Sprintf("total=%.2f", total)
}

type LoyaltyPoints struct{}

func (l *LoyaltyPoints) Points(items []LineItem) int {
	base := 0.0
	for _, item := range items {
		base += item.Price * float64(item.Qty)
	}
	withTax := base + base*0.21
	return int(withTax / 10)
}

func DemoShotgun(items []LineItem) (float64, float64) {
	calc := PriceCalculator{}.TotalWithTax(items)
	inv := InvoiceService{}.CreateTotal(items)
	return calc, inv
}
